planner/core/exhaust_physical_plans.go

```diff
@@ -1498,10 +1516,72 @@ func (p *LogicalJoin) exhaustPhysicalPlans(prop *property.PhysicalProperty) ([]P
 	return joins, true
 }
 
+func (p *LogicalJoin) tryToGetBroadCastJoin(prop *property.PhysicalProperty) []PhysicalPlan {
+	/// todo remove this restriction after join on new collation is supported in TiFlash
+	if collate.NewCollationEnabled() {
+		return nil
+	}
+	if !prop.IsEmpty() {
+		return nil
+	}
+	if prop.TaskTp != property.RootTaskType && !prop.IsFlashOnlyProp() {
+		return nil
+	}
+
+	if p.JoinType != InnerJoin || len(p.LeftConditions) != 0 || len(p.RightConditions) != 0 || len(p.OtherConditions) != 0 || len(p.EqualConditions) == 0 {
+		return nil
+	}
+
+	if hasPrefer, idx := p.getPreferredBCJLocalIndex(); hasPrefer {
+		return p.tryToGetBroadCastJoinByPreferGlobalIdx(prop, 1-idx)
+	}
+	results := p.tryToGetBroadCastJoinByPreferGlobalIdx(prop, 0)
+	results = append(results, p.tryToGetBroadCastJoinByPreferGlobalIdx(prop, 1)...)
+	return results
+}
+
+func (p *LogicalJoin) tryToGetBroadCastJoinByPreferGlobalIdx(prop *property.PhysicalProperty, preferredGlobalIndex int) []PhysicalPlan {
+	lkeys, rkeys := p.GetJoinKeys()
+	baseJoin := basePhysicalJoin{
+		JoinType:        p.JoinType,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		DefaultValues:   p.DefaultValues,
+		LeftJoinKeys:    lkeys,
+		RightJoinKeys:   rkeys,
+	}
+
+	preferredBuildIndex := 0
+	if p.children[0].statsInfo().Count() > p.children[1].statsInfo().Count() {
+		preferredBuildIndex = 1
+	}
+	baseJoin.InnerChildIdx = preferredBuildIndex
+	childrenReqProps := make([]*property.PhysicalProperty, 2)
+	childrenReqProps[preferredGlobalIndex] = &property.PhysicalProperty{TaskTp: property.CopTiFlashGlobalReadTaskType, ExpectedCnt: math.MaxFloat64}
+	if prop.TaskTp == property.CopTiFlashGlobalReadTaskType {
+		childrenReqProps[1-preferredGlobalIndex] = &property.PhysicalProperty{TaskTp: property.CopTiFlashGlobalReadTaskType, ExpectedCnt: math.MaxFloat64}
+	} else {
+		childrenReqProps[1-preferredGlobalIndex] = &property.PhysicalProperty{TaskTp: property.CopTiFlashLocalReadTaskType, ExpectedCnt: math.MaxFloat64}
+	}
```

Comment:
This `if` is not necessary depending on the prop.TaskTp would be either `CopTiFlashGlobalReadTaskType` or `CopTiFlashLocalReadTaskType`?
`&property.PhysicalProperty{TaskTp: prop.TaskTP, ExpectedCnt: math.MaxFloat64}` is ok?

Reply:
It's not ok when prop.TaskTp is rootTask. When it requires rootTask, its child can return a TiFlashTask , close it and get a root task.